dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+```haskell
+infer (Get p n) = do
+    infer p >>= \case
+      Product ds -> do
+        case findTDecl n ds of
+          Just t -> applyBindings t
+          _      -> die $ ExpectedRecordToHaveField n (Product ds)
+
+      other -> do
+        die $ ExpectedRecord other
+```
+
+For the let-bindings, since both their types and bodies are independent of the declaration itself, we infer them before entering the context. We also check  that types of constuctors in datatypes are well-typed _and_ return an object of type being declared. The type of the binding is the type of the context of the binding.
+```haskell
+infer (Let d b) = do
+    ns <- case d of
+      Val n t b' -> do
+        tb <- infer b'
+        t' <- unified (eval t) tb
+        return [(n, t')]
+
+      Data n args ctors -> do
+        for_ args \(TDecl _ t) -> do
+          infer t
+
+        let k = telescope args Star
+        check k Star
+        nts <- for ctors \case
+          Ctor n' t -> do
+            let t' = telescope args t
+            _ <- withContext
+                ((n, eval k) : [(n'', eval t'') | TDecl n'' t'' <- args])
+              do
+                check (telescope args t) Star
+                ret <- getCtorReturnType t
+                ctorCheckReturnType n args ret
+            return [(n', t')]
+        return $ (n, k) : concat nts
+
+    withContext ns do
+      infer b
+```
+
+In case of mutually-recursive definitions, theor bodies (but not the types!) might depend on each other, so we have to bind the constructors to whatever is present in the declarations and do the inference only _after_ we enter the block context.
+```haskell
+infer (LetRec ds b) = do
+    ns <- concat <$> for ds \case
+      Val n t _ -> do
+        check t Star
+        return [(n, eval t)]
+
+      Data n args ctors -> do
+        nts <- for ctors \(Ctor n' t) -> do
+          return [(n', telescope args t)]
+        return $ (n, telescope args Star) : concat nts
+
+    withContext ns do
+      for_ ds \case
+        Val n _ b' -> do
+          t  <- find n
+          t' <- infer b'
+          unified (eval t) t'
+          return ()
+
+        Data n args ctors -> do
+          for_ ctors \case
+            Ctor _ t -> do
+              withContext [(n'', eval t'') | TDecl n'' t'' <- args] do
+                check (eval (telescope args t)) Star
+                ret <- getCtorReturnType t
+                ctorCheckReturnType n args ret
+              return ()
+
+      infer b
+```
+
+There are other cases, but they are either trivial (literals) or the type is contained inside the program directly ("axioms").
+
+## Conclusion
+
+The presence of the typechecker not only allows you to "prevent programs from crashing", but also - if the type system is powerful enough - to partially shift your work onto compiler, by forcing it to implicity calculate some types and, potentially, generate code from them.
+
```

Comment:
I guess the post requires a further reading section.

And mentioning our post written with Vlad :pled: